board.py

```diff
@@ -0,0 +1,194 @@
+        """ Initialize board-specific variables. """
+
+        self.board = []
+        self.board_height = 6
+        self.board_width = 7  # should not exceed 10, due to self.display
+        self.human_piece = '@'
+        self.computer_piece = 'O'
+        self.open_space = '_'
+        self.pieces = self.human_piece + self.computer_piece
+        self.AI_type = 'random'
+
+        self.AI = Opponent(self.AI_type)
+
+        self.status = 'Ongoing'
+        self.players = ['Human', 'Computer']
+
+        # who goes first. (0 is human)
+        self.player_index = 0  # maybe parameterize this
+
+        # Initialize the board as a grid of blank cells
+        for nrow in range(self.board_height):
+            self.board.append([self.open_space for i in
+                               range(self.board_width)])
+
+    def switch_player(self):
+        self.player_index = 1 - self.player_index
+
+    def display(self):
+        """ Display the game state to the terminal. """
+
+        header = ' ' + ' '.join([str(i) for i in range(self.board_width)])
+        num_dashes = 2 * self.board_width - 1
+
+        print(header)
+        print('+%s+' % ('-' * num_dashes))
+
+        for i in range(len(self.board)):
+            print('|%s|' % ('|'.join(self.board[i])))
+
+        print('+%s+' % ('-' * num_dashes))
+        print(header)
+
+    def column_has_vacancy(self, column):
+        """ If board has an opening in <column>, return (True,r), where r
+        is the row where a new piece in that column would fall.
+
+        Otherwise, return (False,-1)
+        """
+
+        r = self.board_height - 1
+        while r >= 0:
+            if self.board[r][column] == self.open_space:
+                return (True, r)
+            r -= 1
+        return (False, -1)
+
+    def attempt_move(self, column, piece):
+        """ Make a move for <piece> in <column>, if possible.
+        Return True iff it was possible.
+        """
+
+        (success, r) = self.column_has_vacancy(column)
+        if success:
+            self.board[r][column] = piece
+        return success
+
+    def get_nonFull_columns(self):
+        """ Return an ordered list of indices of columns which have at least
+            one vacancy.
+        """
+
+        return [c for c in range(self.board_width) if
+                self.column_has_vacancy(c)[0]]
+
+    def request_human_move(self):
+        """ Request a column selection from the user, which is both:
+            - an integer 0 <= c <= [board_width - 1]
+            - the index of a non-full column
+        """
+
+        def is_nonempty_numeric_str(s):
+            digits = '0123456789'
+            if len(s) == 0:
+                return False
+            for c in s:
+                if c not in digits:
+                    return False
+            return True
+
+        got_valid_request = False
+        while not got_valid_request:
+            in_str = input('Column (0-%d)\n>> ' % (self.board_width - 1))
```

Comment:
`input(...)` will return a string, and `isinstance(a, int)` will return `False` on string representations of numbers, like `"5"`. I'm going to keep the existing logic here.

Reply:
```
if __name__ == "__main__":
    #a = input('wat: ')
    a = input('Column (0-%d)\n>> ' % (5 - 1))
    print(type(a))
    print(isinstance(a, int))
```

when I run the above, then enter 5 or 6 (just the numbers, no quotations or w/e) it prints that it's a type ```int``` and ```True```

Reply:
Hmm I seem to get different behavior. With this tiny script:

    cat foo.py 
    a = input('foo: ')
    print(type(a))
    print(isinstance(a, int))

running it gives me:

    python foo.py 
    foo: 4
    <class 'str'>
    False

This might be a distinction between python2 and python3. I'm trying to write the whole project in python3.

Reply:
Oh whoops sorry. Desktop defaults to 2.7 but laptop is 3.6 so I got my versions mixed up

You're right! Although I do think there is a cleaner way to do it that uses the intended data type better. A few lines below you have the following:

```
got_valid_request = self.attempt_move(int(in_str),self.human_piece)
```

would it make sense to sanitize the in_str a little (clear whitespace and all that), then ```try-except``` where you cast to an int?